keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -532,6 +534,45 @@ class SqlArtifactRepository(
     return vetoes.values.toList()
   }
 
+  /**
+   *  Record that an artifact version should be vetoed (marked as bad) in a target environment
+   *
+   *  @param deliveryConfig config object for looking up artifact information by artifact reference
+   *  @param veto information about the artifact version to be vetoed and the target environment to veto in
+   *  @param force if true, record the version as vetoed even if it is marked as the promotion reference
+   *               (automated rollback target) of a previously vetoed version
+   *
+   *  @return true on success
+   *
+   * Preconditions for success:
+   *
+   *  1. [veto.version] is not currently pinned in the target environment
+   *  2. There exists a record R in the environment_artifacts_versions table such that:
+   *     a. R references the artifact version to veto and the target environment encoded in [veto]
+   *     b. R.promotion_reference is NULL, or [force] is true
+   *
+   *  Note: 2b is a precondition to avoid cascading veto-triggered deployments. If R contains a promotion reference, then
+   *  [veto.version] was originally deployed as a result of another artifact version being vetoed (see postcondition 3a).
+   *
+   *
```

Comment:
Here's a dumb question. What's the difference between a "Pre" and "post" condition in this context?

Reply:
Preconditions are predicates which have to be true _before_ you call the function. 

Postconditions are predicates which are supposed to be true _after_ the function completes. 

If the preconditions are true before the function is invoked, and the code is correct, then the postconditions will be true once the function completes.

Basically, if you call the function and the preconditions aren't true, then you can blame the person who wrote the code that calls the function.

If you call the function, and the preconditions are true, but the postconditions aren't true, then you can blame the person who implemented the function.

You can think of the argument types in a Kotlin function declaration as an example of a precondition, and the return type as an example of a postcondition. (In those cases, the preconditions and postconditions are enforced by the compiler).